### ISSUE TITLE
electron-builder includes @types/yargs dependency, TypeScript has type errors on it

### ISSUE THREAD

**Issue by trusktr** [electron-userland/electron-builder#5446]
* **Version**:  22.9.1
* **Electron Version**: N/A
* **Electron Type (current, beta, nightly)**: N/A
* **Target**:  N/A

My project indirectly has `@types/yargs`, due to it being a dependency of `electron-builder`:

```
❯ npm ls @types/yargs
nunisynth@1.0.0 /home/trusktr/src/project-alpha
└─┬ electron-builder@22.9.1
  └── @types/yargs@15.0.10 
```

And TypeScript complains about it:

```
❯ npx tsc
node_modules/@types/yargs/index.d.ts:28:10 - error TS2305: Module '"../../yargs-parser/build/index.cjs"' has no exported member 'DetailedArguments'.

28 import { DetailedArguments, Configuration } from 'yargs-parser';
            ~~~~~~~~~~~~~~~~~

node_modules/@types/yargs/index.d.ts:28:29 - error TS2305: Module '"../../yargs-parser/build/index.cjs"' has no exported member 'Configuration'.

28 import { DetailedArguments, Configuration } from 'yargs-parser';
                               ~~~~~~~~~~~~~


Found 2 errors.
```

Any ideas what's going on with it?

**Comment by trusktr** [comment]
For now, I put `"skipLibCheck": true` in my `tsconfig.json`, but this doesn't seem ideal to be silencing other potential errors.

**Comment by davidje13** [comment]
I have encountered the same problem in a different project. Searching for the error message led me here.

What I found might be useful to you. In short: it looks like `yargs-parser@19` accidentally included typescript `.d.ts` files. They've since been removed in `yargs-parser@20`. I'm guessing something else in your dependencies is pulling in `yargs-parser@19`. The fix is to update it to version 20 (or ask the relevant maintainer to do so). In my case, that project was [neutrino](https://github.com/neutrinojs/neutrino/pull/1619).

More detail:

Annoyingly this leads to a very complicated interaction with NPM's package hoisting behaviour which doesn't know that `yargs-parser` and `@types/yargs-parser` have any sort of link, whereas TypeScript considers definitions included in `yargs-parser` more important than those in `@types/yargs-parser`. If `@types/yargs` finds `yargs-parser` when looking up imports, there's _no guarantee_ that it will be the correct version (since it doesn't have a direct dependency on it, since it's expecting to use `@types/yargs-parser` instead). If it turns out to be version 19, this error happens.

It's possible to make this all work without changing dependencies: if you install packages in the correct order and keep a `package-lock.json`, you can make sure the `yargs-parser@19` isn't hoisted to the root level `node_modules` (some other version will be which doesn't have definitions included). Of course that's pretty fragile so I wouldn't recommend it. Best fix is for everything to update to version 20.

**Comment by ihor-protsenko-theraven** [comment]
I have the same problem 

`node_modules/builder-util-runtime/out/httpExecutor.d.ts:2:10 - error TS2305: Module '"crypto"' has no exported member 'BinaryToTextEncoding'.

2 import { BinaryToTextEncoding } from "crypto";
           ~~~~~~~~~~~~~~~~~~~~

`

**Comment by mmaietta** [comment]
@ihor-protsenko-theraven what version of electron-builder are you referring to. 22.10.5? 22.11.4+?

**Comment by ihor-protsenko-theraven** [comment]
I used 22.10.5,  
`"skipLibCheck": true` is work for me

**Comment by davidje13** [comment]
That error has nothing to do with this issue; the only similarity is that it's of the type "X has no exported member Y", but the module is entirely different ("crypto" != "../../yargs-parser/build/index.cjs")

It looks like you just have an out-of-date version of the Node TypeScript definitions installed; [that alias was added 5 months ago](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/crypto.d.ts#L166), so anything older won't have it. Update your [@types/node](https://www.npmjs.com/package/@types/node) dependency.

**Comment by davidje13** [comment]
Yeah it was never an issue in this project; it was an issue with a yargs update which caused problems here because of some very subtle quirks of typescript & npm (see my earlier post if you want the gory details). In fact I encountered it in a non-electron project; I only posted the description here for the OP and future readers to understand what's going on & how to fix it.

This can be closed. Nothing to fix.
